ros/src/waypoint_updater/waypnt_updater.py

```diff
@@ -188,19 +193,40 @@ def set_waypoint_velocity(self, waypointlist, waypoint, velocity):
     def closest_waypoint(self):
         # TODO - use local search first of final_waypoints sent out last
         # iteration
-        dist = 1000000  # maybe should use max
-        closest = 0
-
         def distance_lambda(a, b): return math.sqrt(
             (a.x-b.x)**2 + (a.y-b.y)**2)
-        for i in range(len(self.waypoints)):
+        if self.final_waypoints:
+            dist = distance_lambda(self.final_waypoints[0].pose.pose.position,
+                                   self.pose.position)
+            for i in range(1, len(self.final_waypoints)):
+                tmpdist = distance_lambda(self.final_waypoints[i].pose.pose.position,
+                                          self.pose.position)
+                if tmpdist < dist:
+                    dist = tmpdist
+                else: 
+                    # distance is starting to get larger so look at 
+                    # last position
+                    if abs(dist-self.last_search_distance) < 5.0:
```

Comment:
This is an interesting check condition.  What kind of situation were you thinking about that would violate this, and how did you choose 5.0 as the threshold?

It seems like this only gets violated if the distance jumped a large distance from the last cycle, over 5m/0.1sec = 50m/s = 112 mph driving perpendicular to the waypoints, so this condition should always by true under normal communication situations.

However, one condition that we also want to fall through to a full search is if the car turns around or goes backwards on the track (or reversing in Carla) behind the last set of final_waypoints.  Since this if condition is still satisfied in this case, the self.last_search_distance is updated to follow the car going backwards farther away and the final_waypoints stay at the last index and doesn't update to move with the car backwards.  You can see this in the simulator if you turn around and drive backward on the track.

Maybe we can first check if it got here on i==1 and if so, break the forward local search loop and start checking a reverse local search incrementally backwards in self.waypoints starting from index self.final_waypoints_start_ptr and decrementing to see if the distance decreases.  If this distance also increases, then final_waypoints[0] is still the closest waypoint.  This method doesn't have a way to fall back to a full search though so maybe there's a condition that would cause a it to get stuck in a local minimum or if there is some localization noise that bounces the pose around or something.

Reply:
I intended to calculate distance travelled since last check by multiplying time_difference * velocity, but stubbed 5.0 in as a reasonable distance that we wouldn't exceed during first test.  I will add the backwards check.  I also need to add a range check for when we loop around the track.

Reply:
I see what you mean for the distance validation, that's a good idea to implement later.

The back search works well, nice!  Should it also have the same distance validation check later?  Also, is the self.back_search flag just for debugging?